Manage lifetime of MiniSat and Glucose solver pointers using unique_ptr

thomasspriggs opened the pull request:
Manage lifetime of MiniSat solver pointer using `unique_ptr`. This has the benefit of allowing the use of a default destructor rather than a hand written destructor. Note that `satcheck_minisat2_baset` was already non-copyable because its `solver_hardness` field was non-copyable. I have initially raised a PR following this approach for the MiniSat solver only, so I can check whether other maintainers see this as a reasonable approach. If this meets with approval then similar changes could be made for other solvers as well.
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

thomasspriggs committed:
Inherit constructor of `satcheck_minisat2_baset`

This avoids having definitions of the constructors of
`satcheck_minisat_simplifiert` and `satcheck_minisat_no_simplifiert`
which were implemented the same way, by having a single template
constructor instead. This will simplify subsequent changes to the
constructor, because they will only need to be made in one place.

thomasspriggs committed:
Declare destructor of `satcheck_minisat2_baset` using `override`

Because this matches the current coding standards, removes a warning
from clang-tidy and points out that this is overriding a base class
destructor.

thomasspriggs committed:
Manage lifetime of MiniSat solver pointer using `unique_ptr`

This has the benefit of allowing the use of a default destructor rather
than a hand written destructor. Note that `satcheck_minisat2_baset` was
already non-copyable because its `solver_hardness` field was
non-copyable.

The default implementation of the destructor is in the `.cpp` rather
than the header file in order to avoid recompiling it every time the
header file is included.

NlightNFotis reviewed (approved):
LGTM.

tautschnig commented:
Thank you @thomasspriggs, this is so much better than #6070! I had made a brief attempt using `unique_ptr`, but it didn't occur to me that I (still) had to place the destructor in the .cpp file. Could you please do the same for Glucose?

thomasspriggs committed:
Inherit constructor of `satcheck_glucose_baset`

This avoids having definitions of the constructors of
`satcheck_glucose_simplifiert` and `satcheck_glucose_no_simplifiert`
which were implemented the same way, by having a single template
constructor instead. This will simplify subsequent changes to the
constructor, because they will only need to be made in one place.

thomasspriggs committed:
Declare destructor of `satcheck_glucose_baset` using `override`

Because this matches the current coding standards, removes a warning
from clang-tidy and points out that this is overriding a base class
destructor.

thomasspriggs committed:
Manage lifetime of Glucose solver pointer using `unique_ptr`

This has the benefit of allowing the use of a default destructor rather
than a hand written destructor. Note that `satcheck_glucose_baset` was
already non-copyable because its `solver_hardness` field was
non-copyable.

The default implementation of the destructor is in the `.cpp` rather
than the header file in order to avoid recompiling it every time the
header file is included and so that the declaration of the solver
destructor is available when instantiating the destructor template of
`unique_ptr`.

thomasspriggs commented:
> Thank you @thomasspriggs, this is so much better than #6070! I had made a brief attempt using `unique_ptr`, but it didn't occur to me that I (still) had to place the destructor in the .cpp file. Could you please do the same for Glucose?

@tautschnig  - Done.
